### DISCUSSION TITLE
Neuromancer models  performance:  comparisons, ideas and suggestions

### DISCUSSION THREAD

**Discussion by mariaade26** [pnnl/neuromancer#240] [Q&A] [answered]
Hello everyone, 

we are currently working with NeuroMANCER to develop, train, and evaluate the performance of various modeling approaches, such as NODE, NSSM, and UDE. These models are applied to learn and predict the dynamics of a single-zone building, which is described by a 1R1C Ordinary Differential Equation (ODE). 

To generate the training data, we did not use NeuroMANCER’s built-in modelSystem simulator. Instead, we created our own synthetic dataset by numerically solving (with odeint) the governing ODE, as demonstrated in the attached notebook. 
The generated data were then used to carry out the following tasks: 

- Parameter estimation of the differential equation, specifically for the thermal resistance (R) and capacitance (C); 
- Training of the non autonomous NSSM model; 
- Training of the non autonomous NODE model; 
- Training of the non autonomous UDE model with full knowledge of the equation structure but without known parameters, requiring parameter estimation;
- Training of the non autonomous UDE model with full knowledge of both the system structure and parameters, thus no parameter estimation was needed. 
- Training of a MLP model; 
- Training of a LSTM model. 

In addition, we also tried to develop an hybrid Universal State Space Model (USSM), combining the UDE and the NSSM features. 

For parameter estimation, we introduced RLogger, a custom extension of the BasicLogger class, designed to track and store the values of the estimated parameters R and C throughout training. This allowed us to monitor the convergence behavior and evaluate the performance of the estimation technique. The results confirmed that the method was effective, as the estimated values of R and C closely matched the true parameters and remained stable during the final 200 epochs of training. 
We introduced the non-autonomous UDE, because we added the dependency of exogenous input to the UDE model and, additionally, we implemented an LSTM block, which was not present. 
For each model, training was performed using the first 2400 samples of the dataset, while the following 500 samples were reserved for testing and evaluation.
We compared the results and the best one seems to be the UDE with the parameters estimation.   

We would like to ask you the following questions: 

1. For the NODE and NSSM models, the training process was straightforward and aligned well with the provided examples. However, when working with the non-autonomous UDE, we observed significantly poor performance—both in terms of prediction accuracy and parameter estimation—when training with normalized input data. To address this, we decided to train the UDE using unnormalized data, which led to satisfactory results. This approach was feasible in our case because all variables happened to have similar magnitudes, minimizing scale-related issues. 
 However, in scenarios where the data variables differ significantly in scale, this workaround would no longer be viable. In such cases, how can we effectively handle parameter estimation and model training under proper normalization? 
  
2. As shown in the graph, even when requires_grad=False is set for the UDE model, we still observe slight variations in the estimated parameters R and C, although these changes are not large. Is it possible to completely prevent any variation in these parameters when requires_grad=False is specified? 

3. We attempted to construct the USSM architecture, but the current implementation does not incorporate any neural network components.  We tried expressing the full matrix form of the system explicitly—for example, defining A=[1/RC] and B=[1/RC ; 1/C ]. However, this approach lacks any neural parameterization, meaning the only learnable components were the matrices A and B. Because of this, we could not set requires_grad=False on these parameters, because doing so would leave the model with nothing to train. 
 We're unsure how to proceed in incorporating both physics-based structure and neural flexibility into this setup. Do you have suggestions for implementing a USSM model that allows both structural interpretability and trainable neural components? 

4. In the Trainer() class, we are able to track the training and validation (deviation) losses, but the testing loss is currently not available during training. Has support for tracking test loss been implemented in newer versions? 

Here we show the results tables for parameters and metrics. 

<img width="945" height="696" alt="Screenshot 2025-08-06 124341" src="https://github.com/user-attachments/assets/81244e5c-18d8-48b6-b152-ae26e1a26aac" />

Further detail can be found in the attached notebook (link below)

https://github.com/mariaade26/NeuromancerModels/blob/main/SysId_ModelsComparison.ipynb

We hope to clarify our remaining doubts soon, and we believe that the presented notebook could serve as a valuable additional NeuroMANCER usage tutorial, complementing the existing examples and documentation. 

Best regards!

**Comment by drgona** [comment, marked as answer]
Hi @mariaade26,

Thanks for putting together such a thorough and detailed notebook! Your model comparison would be an excellent contribution to the neuromancer tutorials and I will be happy to integrate it in the next release.

Comments on the questions you raised:

1. **Handling Normalization in UDEs** - you are right, UDEs with explicit physics structure don't work well with normalized data because the original scale of the physically meaningful variables is distorted.  Few things to try:
- Normalization-aware parameter estimation: If the UDE includes normalized variables, you may need to adjust the parameter learning, e.g., estimate R and C in transformed units or parameterize them relative to the normalization.
- Use latent UDE models with encoder-decoder blocks that essentially learn to denormalize and normalize the data. This allows for greater modeling flexibility but can lose physical interpretability. Make these encoded-decoder maps invertible or affine to improve interpretability
- add explicit non-trainable denormalization and normalization layers in the computational graph to differentiate through - this will allow the model to learn on loss functions in the normalized space but still keep the physical interpretability of the UDE layer
- Employ per-variable normalization: Normalize each feature independently and consider scaling back predictions to original units. This helps the model handle differing variable magnitudes gracefully.

2. **Fixed parameters** - Even if requires_grad=False, parameters might still change if you’re accidentally updating them via optimizer inclusion or logging/debug tools. Make sure you don’t include the frozen parameters in the optimizer’s param list; even with requires_grad=False, an optimizer state can still nudge them. As a safeguard, you might introduce a callback or manual step that reassigns frozen parameter values after each epoch.

3. **Hybrid USSM Architecture** - you have multiple options here to combine a traditional RC network model with neural components. One way to do this is to augment the linear RC net part with neural predictor compensating for modeling errors or disturbances. For example you can keep the A and B matrix to be defined by the RC net and add another neural component learning the mapping of lets say ambient temperature to states. An example of related approaches is shown in this [paper](https://www.sciencedirect.com/science/article/pii/S030626192300435X), or this [one](https://www.sciencedirect.com/science/article/pii/S2666546824000120). There are also other papers you can search with similar ideas.

4. **Trainer** does not print test set loss, but you could pass your “test” set as dev_data if you just want it logged, or modify the trainer to add the print for the test set.

Again, great notebook, and I will be happy to help to integrate it into the main library!

Best,
Jan
